File: code/2352.equal-row-and-column-pairs.php
<?php
/*
 * @lc app=leetcode id=2352 lang=php
 *
 * [2352] Equal Row and Column Pairs
 */

// @lc code=start
class Solution {

    /**
     * @param Integer[][] $grid
     * @return Integer
     */
    function equalPairs($grid) {
        // my solution
        $reverse = [];
        for ($i = 0; $i < count($grid); $i++) {
            for ($j = 0; $j < count($grid); $j++) {
                $reverse[$j][] = $grid[$i][$j];
            }
        }

        $count = 0;
        foreach ($grid as $row) {
            foreach ($reverse as $rrow) {
                if ($rrow == $row) $count++;
            }
        }
        return $count;

        // e.g.hash map
        // $mapRow = [];
        // $mapCol = [];
        // for ($i = 0; $i < count($grid); $i++) {
        //     $colKey = [];
        //     for ($j = 0; $j < count($grid); $j++) {
        //         $colKey[] = $grid[$j][$i]; 
        //     }
        //     $colKey = implode('.', $colKey);
        //     $rowKey = implode('.', $grid[$i]);
        //     $mapCol[$colKey] = ($mapCol[$colKey] ?? 0) + 1;
        //     $mapRow[$rowKey] = ($mapRow[$rowKey] ?? 0) + 1;
        // }

        // $result = 0;
        // foreach ($mapRow as $k => $v) {
        //     if (!isset($mapCol[$k])) {
        //         continue;
        //     }

        //     $result += $v * $mapCol[$k];
        // }

        // return $result;
    }
}
// @lc code=end

File: code/2352.equal-row-and-column-pairs.go
package code

import "strconv"

/*
 * @lc app=leetcode id=2352 lang=golang
 *
 * [2352] Equal Row and Column Pairs
 */

// @lc code=start
func equalPairs(grid [][]int) int {
	length := len(grid)

	rows := make(map[string]int)
	columns := make(map[string]int)

	for i := 0; i < length; i++ {
		row := ""
		column := ""
		for j := 0; j < length; j++ {
			row += strconv.Itoa(grid[i][j]) + "a"
			column += strconv.Itoa(grid[j][i]) + "a"
		}

		rows[row]++
		columns[column]++
	}

	result := 0
	for k, v := range rows {
		if columns[k] > 0 {
			result += v * columns[k]
		}
	}

	return result
}

// @lc code=end
